### DISCUSSION TITLE
Attention mechanism in HypergraphConv

### DISCUSSION THREAD

**Discussion by 1zzt** [pyg-team/pytorch_geometric#7582] [Q&A] [answered]
HypergraphConv first aggregates node features into hyperedges, during which attention weights are applied to evaluate the importance of different nodes in hyperedges.
So should the calculation of alpha be:
alpha = softmax(alpha, hyperedge_index[1], num_nodes=x.size(0)) 
instead of origin code：alpha = softmax(alpha, hyperedge_index[0], num_nodes=x.size(0)) ?

and aggregates hyperedge features into nodes,
So should the calculation of alpha be:
alpha = softmax(alpha, hyperedge_index[0], num_nodes=x.size(0))
instead of origin code：alpha = softmax(alpha, hyperedge_index[1], num_nodes=x.size(0)) ?

Should there be a change in the calculation of alpha?

**Comment by rusty1s** [comment, marked as answer]
As far as I understand the paper, this is correct, as the attention module computes attention for each node belonging to different sets of hyperedges, not attention across all nodes belonging to the same hyperedge.

**Reply by 1zzt** [reply]
Sorry, I didn't understand what you meant, please explain in detail.
The paper says: "For a given vertex xi and its associated hyperedge xj, the attentional score is"
If I want the attentional module to calculate the importance of different nodes in the hyperedge for that hyperedge, should I change the code to: alpha = softmax(alpha, hyperedge_index[1], num_nodes=x.size(0))
Also, if I want the attention module to calculate the importance of different hyperedges to the associated node, should I change the code to: alpha = softmax(alpha, hyperedge_index[0], num_nodes=x.size(0))

Why is this part, not distinguished in the original pyg code? And, should I modify it myself to achieve my purpose? Is my modification correct? I'm confused.
![1686879214180](https://github.com/pyg-team/pytorch_geometric/assets/22312961/aad97952-b419-4212-90ea-da7b24f88ffa)
This is my code：
![image](https://github.com/pyg-team/pytorch_geometric/assets/22312961/fa9528ec-1e88-46a5-8f28-b84e6dd38650)

**Reply by rusty1s** [reply]
> If I want the attentional module to calculate the importance of different nodes in the hyperedge for that hyperedge, should I change the code to: alpha = softmax(alpha, hyperedge_index[1], num_nodes=x.size(0))
Also, if I want the attention module to calculate the importance of different hyperedges to the associated node, should I change the code to: alpha = softmax(alpha, hyperedge_index[0], num_nodes=x.size(0))

Yes, exactly. I feel we can support both options in `HypergraphConv`. Let me give it a try.

**Reply by rusty1s** [reply]
https://github.com/pyg-team/pytorch_geometric/pull/7601

**Reply by 1zzt** [reply]
Thank you for your patient answer!
